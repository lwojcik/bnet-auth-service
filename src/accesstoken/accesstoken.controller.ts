import { Controller, Get, Query } from '@nestjs/common';
import {
  ApiOkResponse,
  ApiOperation,
  ApiTags,
  ApiTooManyRequestsResponse,
  ApiUnauthorizedResponse,
} from '@nestjs/swagger';
import { TooManyRequestsError } from '../common/dto/too-many-requests-error.dto';
import { ApiResponse } from '../common/types';
import { LoggerService } from '../logger/logger.service';
import { AccessTokenService } from './accesstoken.service';
import { AccessTokenError } from './dto/access-token-error.dto';
import { AccessTokenObject } from './dto/access-token-object.dto';
import { GetAccessTokenDto } from './dto/get-access-token.dto';

@ApiTags('accesstoken')
@Controller('accesstoken')
export class AccessTokenController {
  constructor(
    private readonly accessTokenService: AccessTokenService,
    private readonly logger: LoggerService
  ) {
    this.logger.setLoggedClass(AccessTokenController.name);
  }

  @Get()
  @ApiOkResponse({
    description:
      'Successful response. Note that in case of Battle.net errors the application will still return status code 200 and AccessTokenError response.',
    type: AccessTokenObject,
  })
  @ApiUnauthorizedResponse({
    description: ApiResponse.unauthorized,
    type: AccessTokenError,
  })
  @ApiTooManyRequestsResponse({
    description: ApiResponse.tooManyRequests,
    type: TooManyRequestsError,
  })
  @ApiOperation({
    summary: 'Get Battle.net access token for supplied client credentials',
  })
  getAccessToken(
    @Query() getAccessTokenDto: GetAccessTokenDto
  ): Promise<AccessTokenObject | AccessTokenError> {
    this.logger.setLoggedMethod(this.getAccessToken.name, getAccessTokenDto);
    this.logger.debug();
    return this.accessTokenService.getAccessToken(getAccessTokenDto);
  }
}
