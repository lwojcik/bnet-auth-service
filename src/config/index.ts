export * from './app.config';
export * from './auth.config';
export * from './battlenet.config';
export * from './endpoints.config';
export * from './redis.config';
export * from './throttle.config';
