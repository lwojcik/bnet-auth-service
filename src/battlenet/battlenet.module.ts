import { Module } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import { LoggerModule } from '../logger/logger.module';
import { battleNetConfig } from '../config';
import { BattleNetService } from './battlenet.service';

@Module({
  imports: [ConfigModule.forFeature(battleNetConfig), LoggerModule],
  providers: [BattleNetService],
  exports: [BattleNetService],
})
export class BattleNetModule {}
