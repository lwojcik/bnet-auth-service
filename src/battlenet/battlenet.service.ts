import { Inject, Injectable } from '@nestjs/common';
import { ConfigType } from '@nestjs/config';
import { BlizzAPI, RegionName } from 'blizzapi';
import { ApiErrorCode } from '../common/types';
import { battleNetConfig } from '../config';
import { LoggerService } from '../logger/logger.service';

@Injectable()
export class BattleNetService {
  private readonly blizzApi: BlizzAPI;

  constructor(
    @Inject(battleNetConfig.KEY)
    private readonly bnetConfig: ConfigType<typeof battleNetConfig>,
    private readonly logger: LoggerService
  ) {
    this.blizzApi = new BlizzAPI({
      region: this.bnetConfig.region as RegionName,
      clientId: this.bnetConfig.clientId,
      clientSecret: this.bnetConfig.clientSecret,
    });
    this.logger.setLoggedClass(BattleNetService.name);
  }

  async getAccessToken() {
    this.logger.setLoggedMethod(this.getAccessToken.name);
    this.logger.debug();

    try {
      this.logger.debug(
        `Using Battle.net client id: ${this.bnetConfig.clientId}`
      );
      this.logger.debug(
        `Using Battle.net client secret:${this.bnetConfig.clientSecret}`
      );

      const accessToken = await this.blizzApi.getAccessToken();

      this.logger.debug(`Received access token: ${accessToken}`);

      return accessToken;
    } catch (error) {
      this.logger.error(error);

      return {
        error: ApiErrorCode.bnetApiError,
        statusCode: error.response.status,
        message: error.message,
      };
    }
  }
}
