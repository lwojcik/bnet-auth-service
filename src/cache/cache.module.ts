import { RedisModule } from '@liaoliaots/nestjs-redis';
import { Module } from '@nestjs/common';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { redisConfig } from '../config';
import { REDIS } from '../common/constants';
import { CacheService } from './cache.service';
import { LoggerModule } from '../logger/logger.module';

@Module({
  imports: [
    ConfigModule.forFeature(redisConfig),
    LoggerModule,
    RedisModule.forRootAsync({
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: (configService: ConfigService) => ({
        readyLog: true,
        config: {
          host: configService.get(REDIS.host),
          password: configService.get(REDIS.password),
          port: configService.get(REDIS.port),
          db: configService.get(REDIS.db),
          keyPrefix: configService.get(REDIS.keyPrefix),
        },
      }),
    }),
  ],
  providers: [CacheService],
  exports: [CacheService],
})
export class CacheModule {}
