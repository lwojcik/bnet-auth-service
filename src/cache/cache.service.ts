import { RedisService } from '@liaoliaots/nestjs-redis';
import { Inject, Injectable } from '@nestjs/common';
import { ConfigType } from '@nestjs/config';
import Redis from 'ioredis';
import { LoggerService } from '../logger/logger.service';
import { redisConfig } from '../config';
import { REDIS } from '../common/constants';

@Injectable()
export class CacheService {
  private readonly cache: Redis;

  private readonly cacheKey: string;

  constructor(
    private readonly redisService: RedisService,
    @Inject(redisConfig.KEY)
    private readonly redisConf: ConfigType<typeof redisConfig>,
    private readonly logger: LoggerService
  ) {
    this.cache = this.redisService.getClient();
    this.cacheKey = `:${this.redisConf.keyName}`;
    this.logger.setLoggedClass(CacheService.name);
  }

  saveAccessToken(accessToken: string) {
    this.logger.setLoggedMethod(this.saveAccessToken.name, accessToken);
    this.logger.debug();

    if (!this.redisConf.enable) {
      this.logger.debug(`${REDIS.enable} set to false - access key not saved`);
    } else {
      this.logger.debug(
        `Using Redis key: ${this.redisConf.keyPrefix}${this.cacheKey}`
      );

      this.logger.debug(`Set TTL to ${this.redisConf.ttlSecs} seconds`);

      this.cache.set(this.cacheKey, accessToken, 'EX', this.redisConf.ttlSecs);
    }
  }

  async getAccessToken() {
    this.logger.setLoggedMethod(this.getAccessToken.name);
    this.logger.debug();

    if (!this.redisConf.enable) {
      this.logger.debug(
        `${REDIS.enable} set to false - returning 'null' as access key`
      );
      return null;
    }

    this.logger.debug(
      `Using Redis key: ${this.redisConf.keyPrefix}${this.cacheKey}`
    );

    const accessToken = await this.cache.get(this.cacheKey);

    this.logger.debug(`Received access token: ${accessToken}`);

    return accessToken;
  }
}
