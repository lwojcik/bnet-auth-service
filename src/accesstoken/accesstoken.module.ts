import { Module } from '@nestjs/common';
import { RequestContextModule } from 'nestjs-request-context';
import { ConfigModule } from '@nestjs/config';
import { CacheModule } from '../cache/cache.module';
import { BattleNetModule } from '../battlenet/battlenet.module';
import { AccessTokenController } from './accesstoken.controller';
import { AccessTokenService } from './accesstoken.service';
import { LoggerModule } from '../logger/logger.module';
import { redisConfig } from '../config';

@Module({
  imports: [
    BattleNetModule,
    RequestContextModule,
    ConfigModule.forFeature(redisConfig),
    CacheModule,
    LoggerModule,
  ],
  controllers: [AccessTokenController],
  providers: [AccessTokenService],
})
export class AccessTokenModule {}
