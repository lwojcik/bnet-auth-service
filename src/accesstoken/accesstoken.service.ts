import { Inject, Injectable } from '@nestjs/common';
import { RequestContext } from 'nestjs-request-context';
import { ConfigType } from '@nestjs/config';
import { Source } from '../common/types';
import { BattleNetService } from '../battlenet/battlenet.service';
import { GetAccessTokenDto } from './dto/get-access-token.dto';
import { CacheService } from '../cache/cache.service';
import { LoggerService } from '../logger/logger.service';
import { AccessTokenError } from './dto/access-token-error.dto';
import { AccessTokenObject } from './dto/access-token-object.dto';
import { redisConfig } from '../config';

@Injectable()
export class AccessTokenService {
  constructor(
    private readonly battleNetService: BattleNetService,
    private readonly cacheService: CacheService,
    @Inject(redisConfig.KEY)
    private readonly redisConf: ConfigType<typeof redisConfig>,
    private readonly logger: LoggerService
  ) {
    this.logger.setLoggedClass(AccessTokenService.name);
  }

  private async getAccessTokenFromBattleNet() {
    this.logger.setLoggedMethod(this.getAccessTokenFromBattleNet.name);

    this.logger.debug();

    const accessToken = await this.battleNetService.getAccessToken();

    if ((accessToken as AccessTokenError).error) {
      this.logger.debug('Received access token error');
      this.logger.error((accessToken as AccessTokenError).error);

      return {
        ...(accessToken as AccessTokenError),
        id: RequestContext.currentContext.req.id,
        statusCode: (accessToken as AccessTokenError).statusCode,
      };
    }

    return accessToken;
  }

  private async getAccessTokenFromCache(): Promise<string> {
    this.logger.setLoggedMethod(this.getAccessTokenFromCache.name);

    const accessToken = await this.cacheService.getAccessToken();

    this.logger.debug(`Received access token: ${accessToken}`);

    return accessToken;
  }

  private cacheAccessToken(accessToken: string) {
    this.logger.setLoggedMethod(this.cacheAccessToken.name, accessToken);
    this.logger.debug();

    return this.cacheService.saveAccessToken(accessToken);
  }

  async getAccessToken(
    getAccessTokenDto: GetAccessTokenDto
  ): Promise<AccessTokenObject | AccessTokenError> {
    this.logger.setLoggedMethod(this.getAccessToken.name, getAccessTokenDto);
    this.logger.debug();

    const { refresh } = getAccessTokenDto;

    if (this.redisConf.enable && !refresh) {
      this.logger.debug('Redis cache is enabled');
      this.logger.debug('Checking for cached access token...');

      const accessTokenFromCache = await this.getAccessTokenFromCache();

      if (accessTokenFromCache) {
        this.logger.debug(`Found cached access token ${accessTokenFromCache}`);

        return {
          accessToken: accessTokenFromCache,
          source: Source.cache,
        };
      }

      this.logger.debug('No cached access token found!');
    }

    this.logger.debug('Redis cache is disabled OR refresh was triggered');
    this.logger.debug('Getting access token from Battle.net...');

    const accessTokenFromBattleNet = await this.getAccessTokenFromBattleNet();

    if ((accessTokenFromBattleNet as AccessTokenError).error) {
      this.logger.error('Received access token error!');
      return accessTokenFromBattleNet as AccessTokenError;
    }

    if (this.redisConf.enable) {
      this.logger.debug(`Caching access token: ${accessTokenFromBattleNet}...`);
      this.cacheAccessToken(accessTokenFromBattleNet as string);
    }

    return {
      accessToken: accessTokenFromBattleNet as string,
      source: Source.battlenet,
    };
  }
}
