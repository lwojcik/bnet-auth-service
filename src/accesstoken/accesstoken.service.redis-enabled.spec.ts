import { Test, TestingModule } from '@nestjs/testing';
import { ConfigModule, registerAs } from '@nestjs/config';
import { CacheService } from '../cache/cache.service';
import { BattleNetService } from '../battlenet/battlenet.service';
import { AccessTokenService } from './accesstoken.service';
import { LoggerService } from '../logger/logger.service';

jest.mock(
  'nestjs-request-context',
  jest.fn().mockReturnValue({
    RequestContext: {
      currentContext: {
        req: {
          id: 'sample_mocked_request_id',
        },
      },
    },
  })
);

// eslint-disable-next-line global-require
jest.mock('ioredis', () => require('ioredis-mock'));

describe('AccessTokenService (Cache enabled)', () => {
  let serviceWithWarmCache: AccessTokenService;
  let serviceWithEmptyCache: AccessTokenService;
  let serviceWithBattleNetError: AccessTokenService;

  beforeEach(async () => {
    const moduleFactory: ({
      accessTokenInCache,
      battleNetError,
    }: {
      accessTokenInCache?: string;
      battleNetError?: boolean;
    }) => Promise<TestingModule> = ({ accessTokenInCache, battleNetError }) =>
      Test.createTestingModule({
        imports: [
          ConfigModule.forFeature(
            registerAs('redis', () => ({
              enable: true,
            }))
          ),
        ],
        providers: [
          {
            provide: BattleNetService,
            useValue: {
              getAccessToken: () =>
                battleNetError
                  ? {
                      error: 'Mocked error response',
                      statusCode: 'Mocked error status code',
                    }
                  : 'mocked_access_token_from_battlenet',
            },
          },
          {
            provide: CacheService,
            useValue: {
              getAccessToken: () => accessTokenInCache || null,
              saveAccessToken: jest.fn(),
            },
          },
          {
            provide: LoggerService,
            useValue: {
              debug: jest.fn(),
              log: jest.fn(),
              error: jest.fn(),
              setLoggedClass: jest.fn(),
              setLoggedMethod: jest.fn(),
            },
          },
          AccessTokenService,
        ],
      }).compile();

    serviceWithWarmCache = (
      await moduleFactory({
        accessTokenInCache: 'test_access_token_from_redis_cache',
      })
    ).get<AccessTokenService>(AccessTokenService);

    serviceWithEmptyCache = (await moduleFactory({})).get<AccessTokenService>(
      AccessTokenService
    );

    serviceWithBattleNetError = (
      await moduleFactory({ battleNetError: true })
    ).get<AccessTokenService>(AccessTokenService);
  });

  describe('Access token saved in cache', () => {
    it('should get access token from cache', async () => {
      expect.assertions(1);
      const accessToken = await serviceWithWarmCache.getAccessToken({});
      expect(accessToken).toMatchSnapshot();
    });

    it('should get access token from Battle.net when refresh is true', async () => {
      expect.assertions(1);
      const accessToken = await serviceWithWarmCache.getAccessToken({
        refresh: true,
      });
      expect(accessToken).toMatchSnapshot();
    });
  });

  describe('No access token saved in cache', () => {
    it('should get access token from Battle.net', async () => {
      expect.assertions(1);
      const accessToken = await serviceWithEmptyCache.getAccessToken({});
      expect(accessToken).toMatchSnapshot();
    });

    it('should return access token error if Battle.net API returns an error', async () => {
      expect.assertions(1);
      const accessToken = await serviceWithBattleNetError.getAccessToken({});
      expect(accessToken).toMatchSnapshot();
    });

    it('should get access token from Battle.net if refresh is set to true', async () => {
      expect.assertions(1);
      const accessToken = await serviceWithEmptyCache.getAccessToken({
        refresh: true,
      });
      expect(accessToken).toMatchSnapshot();
    });
  });
});
