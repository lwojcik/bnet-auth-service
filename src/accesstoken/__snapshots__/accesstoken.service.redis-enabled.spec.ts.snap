// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccessTokenService (Cache enabled) Access token saved in cache should get access token from Battle.net when refresh is true 1`] = `
Object {
  "accessToken": "mocked_access_token_from_battlenet",
  "source": "battlenet",
}
`;

exports[`AccessTokenService (Cache enabled) Access token saved in cache should get access token from cache 1`] = `
Object {
  "accessToken": "test_access_token_from_redis_cache",
  "source": "cache",
}
`;

exports[`AccessTokenService (Cache enabled) No access token saved in cache should get access token from Battle.net 1`] = `
Object {
  "accessToken": "mocked_access_token_from_battlenet",
  "source": "battlenet",
}
`;

exports[`AccessTokenService (Cache enabled) No access token saved in cache should get access token from Battle.net if refresh is set to true 1`] = `
Object {
  "accessToken": "mocked_access_token_from_battlenet",
  "source": "battlenet",
}
`;

exports[`AccessTokenService (Cache enabled) No access token saved in cache should return access token error if Battle.net API returns an error 1`] = `
Object {
  "error": "Mocked error response",
  "id": "sample_mocked_request_id",
  "statusCode": "Mocked error status code",
}
`;
